src/Query/Group.ts

```diff
@@ -151,13 +162,44 @@ export class Group {
         return [Group.escapeMarkdownCharacters(filename)];
     }
 
-    private static groupByRoot(task: Task): string[] {
-        const path = task.path.replace(/\\/g, '/');
-        const separatorIndex = path.indexOf('/');
-        if (separatorIndex == -1) {
-            return ['/'];
+    private static groupByFn(task: Task, arg?: GroupingArg): string[] {
+        const paramsArgs: [string, any][] = [
```

Comment:
There is some documentation in CONTRIBUTING.md about what to do when adding a new field to `Task`. It would be worth this PR including an update to that, to point people to this new location that would also be worth adding a new field to.

Reply:
Another option for a v1 maybe could be to just pass task/t and some convenient extras not already in the task object

```typescript
const paramsArgs: [string, any][] = [
    ['path', task.path.replace('.md', '')],
    ['root', Group.root(task)],
    ['t', task],
    ['task', task],
];
```

Reply:
I ask because I care a lot about avoiding breaking changes for users - not requiring them to rewrite tasks blocks if at all possible.

In #1249 I tried to explore whether making `Task` objects visible to users - with all its public data - would tie my hands and prevent me from doing future refactorings.

For that reason I started implementing an abstraction for dates, but didn't get very far as a lot of other stuff has cropped up.

I am therefore a little nervous about this PR's exposing of raw dates in this PR, and what restrictions it would mean on future improvements to code maintainability.

And considerably more nervous at the idea of exposing the entire `Task` class.

I am prepared to be persuaded away from that view, if someone can show me ways of exposing the inner details now, and not hampering future development... I just personally can't see a way right now.

The background to all this is that there are a load of missing abstractions that make addition of new features and comprehension of the code harder than I would like. And so I am trying to juggle improving the design with user support and PRs... (It's a nice position to be in, of course...)